handlers/validator.go

```diff
@@ -944,6 +835,148 @@ func hasMultipleWithdrawalCredentials(deposits *types.ValidatorDeposits) bool {
 	return false
 }
 
+func handleValidatorWithNoIndex(w http.ResponseWriter, r *http.Request, pubKey []byte, index uint64, validatorPageData *types.ValidatorPageData, data *types.PageData) {
+	validatorTemplateFiles := append(layoutTemplateFiles,
+		"validator/validator.html",
+		"validator/heading.html",
+		"validator/tables.html",
+		"validator/modals.html",
+		"modals.html",
+		"validator/overview.html",
+		"validator/charts.html",
+		"validator/countdown.html",
+		"components/flashMessage.html",
+		"components/rocket.html")
+	validatorNotFoundTemplateFiles := append(layoutTemplateFiles, "validator/validatornotfound.html")
+	var validatorTemplate = templates.GetTemplate(validatorTemplateFiles...)
+	var validatorNotFoundTemplate = templates.GetTemplate(validatorNotFoundTemplateFiles...)
+
+	var name string
+	err := db.ReaderDb.Get(&name, `SELECT name FROM validator_names WHERE publickey = $1`, pubKey)
+	if err != nil && err != sql.ErrNoRows {
+		logger.Errorf("error getting validator-name from db for pubKey %v: %v", pubKey, err)
+		http.Error(w, "Internal server error", http.StatusInternalServerError)
+		return
+		// err == sql.ErrNoRows -> unnamed
+	} else {
+		validatorPageData.Name = name
+	}
+
+	var pool string
+	err = db.ReaderDb.Get(&pool, `SELECT pool FROM validator_pool WHERE publickey = $1`, pubKey)
+	if err != nil && err != sql.ErrNoRows {
+		logger.Errorf("error getting validator-pool from db for pubKey %v: %v", pubKey, err)
+		http.Error(w, "Internal server error", http.StatusInternalServerError)
+		return
+		// err == sql.ErrNoRows -> (no pool set)
+	} else {
+		if validatorPageData.Name == "" {
+			validatorPageData.Name = fmt.Sprintf("Pool: %s", pool)
+		} else {
+			validatorPageData.Name += fmt.Sprintf(" / Pool: %s", pool)
+		}
+	}
+	deposits, err := db.GetValidatorDeposits(pubKey)
+	if err != nil {
+		logger.Errorf("error getting validator-deposits from db: %v", err)
+	}
+	validatorPageData.DepositsCount = uint64(len(deposits.Eth1Deposits))
+	validatorPageData.ShowMultipleWithdrawalCredentialsWarning = hasMultipleWithdrawalCredentials(deposits)
+	if err != nil || len(deposits.Eth1Deposits) == 0 {
```

Comment:
Interesting, I did not know such validators can exist.

I checked for 51737 but there does not seem to be a problem. Whether you search by index or pubkey you do not enter `handleValidatorWithNoIndex` because a balance for the validator is found in the finalized epochs.

Can a new validator without deposits exist?
If it can I would imagine that like for normal new ones before my fix it would cause an ISE the same way before a balance is in the finalized epochs.
Before the fix the logic was:

- If you search for a validator with a pubkey but it has no **index** and no **Initiated Deposit** then show "validator not found"

Now after the fix it has an additional:

- If you search for a validator and it has an **index** but no **balance** is in the finalized epochs and no **Initiated Deposit** then show "validator not found"

This would make sense to me but what is your opinion? Should something else be shown?

Reply:
According to @Buttaa, these validators are normal for testnets and were created by changing the state. This can't happen on mainnet and shouldn't happen for new validators on testnets. My personal opinion is that since these validators are an exception to the norm it's acceptable to not show them until their balance shows up in a finalized epoch (might happen for new testnets idk).